### PR TITLE
[WIP] Add stochastic sampling, preconditioning, and other improvements to HeunDiscreteScheduler

### PR DESCRIPTION
**Pull request by dg845**
# What does this PR do?

This PR tries to make the `HeunDiscreteScheduler` match Algorithm 2 (the Karras stochastic sampler) of the [EDM paper](https://arxiv.org/pdf/2206.00364.pdf) more closely. The main changes are:

- Supporting stochastic sampling (`gamma > 0`). 
- Adding noise and output preconditioning (see Table 1 of the [EDM paper](https://arxiv.org/pdf/2206.00364.pdf))
- Adding custom `timesteps` support to make it easier to use specific `sigma` schedules

This PR is related to #4481 but modifies the existing `HeunDiscreteScheduler` instead of creating a new scheduler.

image

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@patrickvonplaten
@williamberman
@yiyixuxu

### CODE REVIEW

**Comment by dg845** [comment]
Some notes on the design:

- The increased noise levels `sigma_hat` ($`\hat{t}_i`$ in the original paper) [Line 5 in Alg. 2] are calculated beforehand in `set_timesteps`, since everything to calculate the `gamma` noise factors ($`\gamma_i`$) is available there.
- Noise preconditioning (scaling the timestep input $`\sigma`$ by $c_{\mbox{noise}}(\sigma) = \frac{1}{4}\ln{(\sigma)}$) is applied directly to the `timesteps`, to avoid needing a `scale_timestep_input` method.
- Adding noise for stochastic sampling [Line 6] to get the increased noise sample `sample_hat` ($`\hat{x}_i`$) happens in `scale_model_input`, since we need to give `sample_hat` to the U-Net model for the first order Euler model evaluation. This modifies the `scale_model_input` interface since there's an extra (optional) `generator` argument for reproducibility.
- Input preconditioning (scaling the model input $`x`$ by $`c_{\mbox{in}}(\sigma) = 1 / \sqrt{\sigma^2 + \sigma_{\mbox{data}}^2}`$) happens after potentially adding noise to the model input in `scale_model_input`. I think there might be a discrepancy between the paper and the current `HeunDiscreteScheduler` implementation, where the paper uses the sample _before_ scaling when performing Euler and Heun steps, while the implementation uses the sample _after_ scaling. The test code follows the implementation, e.g. in `HeunDiscreteScheduler.test_full_loop_no_noise`:

https://github.com/huggingface/diffusers/blob/ea311e6989ca8b36b765450f57f13901c03cb0af/tests/schedulers/test_scheduler_heun.py#L55-L61

but perhaps it should be

```python
    for i, t in enumerate(scheduler.timesteps):
        scaled_sample = scheduler.scale_model_input(sample, t)

        model_output = model(scaled_sample, t)

        output = scheduler.step(model_output, t, sample)
        sample = output.prev_sample
```
- Currently, the output preconditioning

$$D_\theta(x; \sigma) = c_{\mbox{skip}}(\sigma)x + c_{\mbox{out}}(\sigma)F_\theta(c_{\mbox{in}}(\sigma)x; c_{\mbox{noise}}(\sigma))$$

happens after getting `pred_original_sample`, but perhaps it makes more sense to consider it its own `prediction_type` (since it modifies how the U-Net $`F_\theta`$ is trained).

**Comment by dg845** [comment]
Also, would someone be able to explain to me/provide a reference for what the `_sigma_to_t` method does? I can't make sense of it.

**Review by patrickvonplaten — COMMENTED** [review]
@yiyixuxu can you take a look here?

**Comment by dg845** [comment]
To expand more on https://github.com/huggingface/diffusers/pull/4930#issuecomment-1711056214, I guess I am having trouble understanding why this works. In the framework of the [Karras paper](https://arxiv.org/pdf/2206.00364.pdf), once we have `sigmas` $`\{\sigma_i\}`$ (e.g. from the Karras sigmas function or the `alpha`/`beta` schedule), the timesteps $`\{t_i\}`$ are determined by $`t_i = \sigma^{-1}(\sigma_i)`$, where $`\sigma(\cdot)`$ specifies a noise schedule from the `sigmas` $`\{\sigma_i\}`$. The EDM model and implementation uses $`\sigma(t) = t = \sigma^{-1}(t)`$ (see Section 3), but I'm having trouble figuring out what $`\sigma(\cdot)`$ function `_sigma_to_t` expresses, and what motivates this choice.

(Also, is there a bug in the `_sigma_to_t` implementation? If I do something like

```python
import numpy as np

sigma_min = 0.002
sigma_max = 80.0
rho = 7.0
num_inference_steps = 10

# Calculate Karras sigmas and log sigmas
ramp = np.linspace(0, 1, num_inference_steps)
min_inv_rho = sigma_min ** (1 / rho)
max_inv_rho = sigma_max ** (1 / rho)
sigmas = (max_inv_rho + ramp * (min_inv_rho - max_inv_rho)) ** rho
log_sigmas = np.log(sigmas)

# Map sigmas back to timesteps
timesteps = []
for i, sigma in enumerate(sigmas):
	log_sigma = np.log(sigma)
	dists = log_sigma - log_sigmas[:, np.newaxis]
	low_idx = np.cumsum((dists >= 0), axis=0).argmax(axis=0).clip(max=log_sigmas.shape[0] - 2)
	high_idx = low_idx + 1
	low = log_sigmas[low_idx]
	high = log_sigmas[high_idx]
	w = (low - log_sigma) / (low - high)
	w = np.clip(w, 0, 1)
	t = (1 - w) * low_idx + w * high_idx
	t = t.reshape(sigma.shape)
	timesteps.append(t)
timesteps = np.array(timesteps)
print(timesteps)
```

I get the output

```python
array([8., 8., 8., 8., 8., 8., 8., 8., 8., 9.])
```

which doesn't seem reasonable. It feels like there should be a `.max` instead of a `.argmax` in the line where we get `low_idx`: since `(dists >= 0)` is non-negative and strictly positive at the tail end [because the Karras `sigmas` are strictly decreasing] when we cast it to a float, `np.cumsum((dists >=0), axis=0)` will be strictly increasing at the end, so the `argmax` will always return the maximum index possible, which will get clipped to `num_inference_steps - 2`. So seems that the timestep will always hover between `num_inference_steps - 2` and `num_inference_steps - 1`.)

**Comment by dg845** [comment]
Gentle ping @yiyixuxu.

**Comment by dg845** [comment]
Gentle ping @yiyixuxu @patrickvonplaten for an initial design review. As a note, the tests in `tests/schedulers/test_scheduler_heun.py` pass (although I haven't added tests for the new functionalities like custom timesteps yet).

**Comment by patrickvonplaten** [comment]
@yiyixuxu can you take a look here?

**Comment by yiyixuxu** [comment]
@dg845 

Am I correct that we only need to move forward with either this PR or https://github.com/huggingface/diffusers/pull/4481? 

can we see some results with this new scheduler (or new variant) please?

**Comment by dg845** [comment]
Hi @yiyixuxu, sorry for the late reply.

> Am I correct that we only need to move forward with either this PR or https://github.com/huggingface/diffusers/pull/4481?

Yes, that is correct. I think this PR would be preferable all things being equal because it modifies an existing scheduler rather than adding a new scheduler. (Note that there is currently a slight change to the `scale_model_input` API in this PR which adds an extra `generator` argument.)

> can we see some results with this new scheduler (or new variant) please?

I will work on this.

**Comment by yiyixuxu** [comment]
@dg845

 I think you have already iterated quiet a bit on this PR https://github.com/huggingface/diffusers/pull/4481 - that's just use the new scheduler on that PR to show some results! we can decide which PR to move forward later :)  - would this be ok?

**Comment by dg845** [comment]
Sounds good :).
